### PR TITLE
ENH: optimize: add inv Hess estimate in BFGS

### LINKED ISSUE #17864: ENH: feature request for initial hessian estimate in scipy bfgs routine
### Is your feature request related to a problem? Please describe.

At the moment `scipy.optimize.minimize(..., method = 'BFGS')` does not accept an initial estimate for the inverse of the Hessian. This would be helpful because

1) it would make the algorithm more efficient when a good estimate is available
2) it allows using information from a previous second-order optimization routine which failed to converge because of the max number of iterations being reached. 

### Describe the solution you'd like.

It is easy to change 

`    I = np.eye(N, dtype=int)
    Hk = I` 

from lines 1373 - 1384 of [bfgs_implementation](https://github.com/scipy/scipy/blob/v1.10.0/scipy/optimize/_optimize.py#L1189-L1315)

to 

`    Hk = Hk_initial_estimate`

where  `Hk_initial_estimate` is a parameter to be passed by the user, and which is by default set to `np.eye(N, dtype = int)`. It can also be easily checked if the initial estimate is positive definite and symmetric.

### Describe alternatives you've considered.

The jax.scipy and tfp bfgs routines hang, hence I have resorted to the scipy implementation and only estimate the function and the gradients with jax on a gpu.  This would also benefit users of jax and tfp.

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by paulestano**
#### Reference issue
Closes scipy#17864.

#### What does this implement/fix?
Add inverse Hessian estimate parameter `hk_init`  in `_minimize_bfgs`.

#### Additional information
`hk_init` is checked to be positive definite and initializes `Hk` if not `None` else `Hk` is set to identity. 
Add unit-test to ensure non positive definite `hk_init` are not allowed.
`hk_init`  is also added to `fmin_bfgs` signature.

### CODE REVIEW

File: scipy/optimize/_optimize.py
Comment (ilayn) [inline scipy/optimize/_optimize.py:275]: You can use `linalg.issymmetric` instead here, to have a bit more performance. But it does an exact comparison of the off-diagonal elements so no tolerances. However if you provide tolerances it falls back to this test you have here.

File: scipy/optimize/_optimize.py
Comment (ilayn) [inline scipy/optimize/_optimize.py:277]: Use `scipy.linalg.cholesky`.

File: scipy/optimize/_optimize.py
Comment (paulestano) [reply scipy/optimize/_optimize.py:275]: Thanks for the suggestion. I was not aware of this function. Will make the change for readability and consistency.

**Comment by dschmitz89** [comment]
Thanks for the PR!

Two comments from my side:

1. We need to test that the new parameter works, not only the error. One possibility could to check that the result of running with and without the parameter is the same for a simple problem. A second possible test would be that with the new parameter the number of iterations is reduced because of the good initial estimate.
2. I think we can give the new argument a more descriptive name: what about `hinv0` or `hinv_init`? `Hk` indicates to me "something related to Hessian" while including "inv" makes it clear what it is about.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Thanks for the PR!

Two comments from my side:

1. We need to test that the new parameter works, not only the error. One possibility could to check that the result of running with and without the parameter is the same for a simple problem. A second possible test would be that with the new parameter the number of iterations is reduced because of the good initial estimate.
2. I think we can give the new argument a more descriptive name: what about `hinv0` or `hinv_init`? `Hk` indicates to me "something related to Hessian" while including "inv" makes it clear what it is about.


A little offtopic: for my PR #18986 I also need the check for (semi) positive definiteness. I think it would make sense to have this as a utils function somewhere in `scipy._lib`. I am not a `linalg` expert, so I am not sure what the most performant check would be though.

**Comment by paulestano** [comment]
Thanks for your review and comments @dschmitz89. I'll reply inline:
> 1. We need to test that the new parameter works, not only the error. One possibility could to check that the result of running with and without the parameter is the same for a simple problem. A second possible test would be that with the new parameter the number of iterations is reduced because of the good initial estimate.

I actually thought about that. I am going to commit a test with `rosen` for your first possibility. If necessary I can also look for examples for the second though I think it's up to the user to choose a 'performant' intial estimate (as long as it is positive definite) 
> 2. I think we can give the new argument a more descriptive name: what about `hinv0` or `hinv_init`? `Hk` indicates to me "something related to Hessian" while including "inv" makes it clear what it is about.

`Hk` is the variable representing the inverse of the hessian estimate in this function. Even though I can't speak for the developer who wrote this code, I'm assuming he followed the naming conventions of Nocedal which defines $H_k=B_k^{-1}$ (where $B_k$ is the Hessian matrix) in [Numerical Optimization (p. 139)](https://link.springer.com/book/10.1007/978-0-387-40065-5). 
I think textbook naming conventions don't always work but I don't think we should make this refactor (if required) in this PR.
However, I agree `hk_init` to be confusing and I suggest to rename it `h0` if that works.

>A little offtopic: for my PR #18986 I also need the check for (semi) positive definiteness. I think it would make sense to have this as a utils function somewhere in `scipy._lib`. I am not a `linalg` expert, so I am not sure what the most performant check would be though.

I am not a `linalg` expert either so I trust @ilayn on that and will switch to `scipy.linalg.issymmetric`. I want to point out that I'm (only) checking for strict positive definiteness, while semi-positive definiteness might be trickier as a cholesky decomposition only exists for strictly positive definite matrices. This may even require a test as part of this PR (i.e. ensuring that semi-positive definite matrices are not allowed as initial inverse Hessian estimate).

**Comment by paulestano** [comment]
Tried to address comments from @ilayn about `_check_positive_definite` and offer tentative solutions to comments from @dschmitz89 about testing and naming. 
I'm happy to change `h0` to something else if this new name does not work.

**Comment by ilayn** [comment]
There is no numerically cheap test for semi-definiteness and I think you don't want to have it here anyways. But I'll leave it to optimize residents. 

I should clarify. What I meant with `issymmetric` is that, you don't want to provide any tolerances to have an exact comparison of off-diagonal elements. If you provide tolerances you are leaving room for almost symmetric arrays which will differ which half you take in the cholesky computations.

**Comment by paulestano** [comment]
Thanks for the clarifications @ilayn. I removed the tolerance parameters from the `issymmetric` call.

File: scipy/optimize/tests/test_optimize.py
Comment (dschmitz89) [inline scipy/optimize/tests/test_optimize.py:239]:
```suggestion
        assert_allclose(res.fun), res_true.fun, atol=1e-6)
```

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Tests look good to me now. After another sweep of the optimize docs, I think an even better name for the initial inverse hessian is `hessinv0` as `hess` is used for the Hessian function. But that's my personal opinion. Let's do a poll via emojis to get an answer from more people. Please react to my post with the respective emoji.

- `hk` 👍 
- `h0`  🚀 
- `hinv0` ❤️ 
- `hessinv0` 👀

**Comment by dschmitz89** [comment]
5 days have passed, let's go with `hess_inv0` then. I think that is even better than `hessinv0` as the [`OptimizeResult`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html) contains `hess_inv`.

**Comment by paulestano** [comment]
Hi @dschmitz89. I just pushed the change to `hess_inv0`. All the initial comments should hopefully be addressed now

**Review by dschmitz89 — APPROVED** [review]

**Comment by dschmitz89** [comment]
Test failures come from cython (see #19132).

**Comment by ilayn** [comment]
@mdhaber @andyfaff could you chime in for the variable name selection and if you agree, probably this is mergable.

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:287]:
```suggestion
        
        
```

Comment (andyfaff) [reply scipy/optimize/_optimize.py:287]: double empty line between functions.

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1312]: State shape that the ndarray should have?

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:286]:
```suggestion
            raise ValueError("'hess_inv0' matrix isn't positive definite.")
```

File: scipy/optimize/tests/test_optimize.py
Comment (andyfaff) [inline scipy/optimize/tests/test_optimize.py:223]:
```suggestion
        # Ensure that BFGS does not accept semi pos. def. initial inverse Hessian estimate.
```

File: scipy/optimize/tests/test_optimize.py
Comment (andyfaff) [inline scipy/optimize/tests/test_optimize.py:215]:
```suggestion
        # Ensure that BFGS does not accept neg. def. initial inverse Hessian estimate.
```

File: scipy/optimize/tests/test_optimize.py
Comment (andyfaff) [inline scipy/optimize/tests/test_optimize.py:216]:
```suggestion
        with pytest.raises(ValueError, match="'hess_inv0' matrix isn't positive definite."):
```

Comment (andyfaff) [reply scipy/optimize/tests/test_optimize.py:216]: Also the other ValueError messages.

File: scipy/optimize/tests/test_optimize.py
Comment (andyfaff) [inline scipy/optimize/tests/test_optimize.py:238]: Check all lines in PR are less than 81 chars.

File: scipy/optimize/tests/test_optimize.py
Comment (andyfaff) [inline scipy/optimize/tests/test_optimize.py:237]: Line length

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1448]: Update docstring to reflect comments above.

**Review by andyfaff — COMMENTED** [review]
I'd like to see a greener CI before merge.

File: scipy/optimize/_optimize.py
Comment (paulestano) [reply scipy/optimize/_optimize.py:1312]: The shape of `x0` is not documented in any `fmin_*` or `_minimize_*` function. Shall I add it at least for `_minimize_bfgs` as the shape of the Hessian will depend on it?

**Comment by paulestano** [comment]
Thanks for your review @andyfaff. My last commit should hopefully address your comments.
I added a `shape (n,)` in `x0` docstrings as I needed it to define`hess_inv0` shape. Let me know if that does not work.

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1312]:
```suggestion
    hess_inv0 : None or ndarray, optional``

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1279]:
```suggestion
    x0 : ndarray
```

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1280]:
```suggestion
        Initial guess, shape (n,)
```

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1313]:
```suggestion
        Initial inverse hessian estimate, shape (n, n). If None (default) then the identity
```

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1449]:
```suggestion
    hess_inv0 : None or ndarray, optional
```

File: scipy/optimize/_optimize.py
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1450]:
```suggestion
        Initial inverse hessian estimate, shape (n, n). If None (default) then the identity
```

**Comment by paulestano** [comment]
I don't know much about meson and Cirrus but it there some caching happening here? In any case I don't think this PR impact `scipy.conftest.py`
```
python dev.py test -j1
💻  ninja -C /tmp/cirrus-ci-build/build -j1
ninja: Entering directory `/tmp/cirrus-ci-build/build'
[1/70] Generating scipy/generate-version with a custom command
[2/70] Generating scipy/special/_dummy_cython_special.pxd with a custom command
[3/67] Generating scipy/linalg/_dummy_cython_blas.pxd with a custom command
[4/31] Generating scipy/stats/_dummy_stats_gen_pyx with a custom command
Build OK
💻  meson install -C build --only-changed
ImportError while loading conftest '/tmp/cirrus-ci-build/build-install/lib/python3/dist-packages/scipy/conftest.py'.
scipy/conftest.py:11: in <module>
    import hypothesis
E   ModuleNotFoundError: No module named 'hypothesis'
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /tmp/cirrus-ci-build/build-install/lib/python3/dist-packages
Running tests for scipy version:1.12.0.dev0+1572.9a43ba7, installed at:/tmp/cirrus-ci-build/build-install/lib/python3/dist-packages/scipy
����
```

**Comment by andyfaff** [comment]
THe issue is that the CI setup isn't installing `hypothesis`, which is a new test dependency. @rgommers, this is the clone issue I was referring to yesterday. For some reason the clone isn't working properly, I'll have to look into it. I think it's mergeable as-is though.

**Comment by rgommers** [comment]
Ah thanks, that explanation makes sense. I guess that's the consequence of doing the git merge with main inside the yml file rather than the CI system doing it before running anything. 

I'd like to de-duplicate these requirements and install them straight from `pyproject.toml` (better than scattering more requirements files around). Let's figure that out in a separate issue?

**Comment by NickAlger** [comment]
Thank you for adding this functionality. This has been a major missing piece of scipy's L-BFGS for a long time now.

**Comment by NickAlger** [comment]
Thanks for adding this functionality. The same would be very useful for L-BFGS. The required change is to have the initial inverse Hessian approximation multiply q on line 474 of the lbfgs file here:
https://github.com/scipy/scipy/blob/3f9a8c80e281e746225092621935b88c1ce68040/scipy/optimize/_lbfgsb_py.py#L474

This corresponds to the line r<-H_k^0 q; in Algorithm 7.4 on page 178 of Nocedal and Wright.

**Comment by dschmitz89** [comment]
@NickAlger : I think your last point is already in discussion in https://github.com/scipy/scipy/issues/18105. Could you join the discussion in that issue? Discussions in merged PRs are difficult to discover. Adding `H_0` for `L-BFGS-B` also sounds useful to me.

**Pull request merged**
